[14.0][IMP] base_exception: Placeholders for exception rule description

raumschmiede-joshuaL opened the pull request:
Hello.

I changed the exception rules to use Jinja in exception rule descriptions.
For that I changed a bit more in the module so that for exceptions that were detected on underlying records (like in `sale_exception`) the text is still correctly added to the exceptions summary on the parent records. I adjusted the tests to cover this.

Useful for us as we have a lot of sales exceptions. Some exceptions check whether any child of a sold BOM product contains valid data. For BOMs with lots of children or BOMs in BOMs it is not helpful enough for the user to see that the 1st SO line has the exception "The product has no weight set".
With my change you can use "The product {{obj.default_code}} has no weight set" as description.


My changes do not break modules that depend on `base_exception`. I tested it on `sale_exception`. The only thing that wouldn't work in `sale_exception` is if you use the placeholders that the exceptions summary for the SO line wouldn't be correctly computed as `sale_exception` defines its own fields on the SO line.
**EDIT 2**: With the latest changes `sale_exception` will break,


~I wasn't sure whether it is better to have Jinja or a separate code field as placeholders. I like it more to write code, normal non-dev users would rather use Jinja, like in mail templates. That's why I added both to make everyone happy.
But for the Jinja stuff I added a dependency to "mail" as this provides already a function that parses Jinja. I saw in another OCA module that it was used like that. I did the same to keep it simple and consistent.
If this new dependency is not good, I could move this into a new module.~

**EDIT 2**: I removed the description code placeholders. Can't say whether they are good for the OCA module. I will add this in a separate module in our repo.


I checked which modules depend on `base_exception`. The ones I found I already adjusted to work with the new field to define which fields contain sub-records for which exceptions must be checked.
So far I found 5 modules. I already changed the code in them, but I would create the PRs to change them later after the discussion here.
**EDIT**: I already added the PR for `sales_exception`: https://github.com/OCA/sale-workflow/pull/3861. Can be used as example or for testing if you want. And for that more changes are needed. 


Everything is still more a draft. I would like to hear your opinion on the changes. Also about naming new defs and fields.


Screenshots:


Jinja in exception description:
<img width="940" height="604" alt="image" src="https://github.com/user-attachments/assets/f482240d-b122-4b33-be1b-d53c876e1382" />
**EDIT 2** Is now `{{...}}` and not `${...}` to avoid new dependency and to have the same as in later Odoo versions.

Exceptions in sales order:
<img width="1534" height="757" alt="image" src="https://github.com/user-attachments/assets/0ed621f2-b14b-446e-98f9-9498cb820025" />

raumschmiede-joshuaL commented:
@mt-software-de, @jbaudoux, @i-vyshnevska

mt-software-de commented on base_exception/models/base_exception.py:
```diff
@@ -21,7 +22,43 @@ class ExceptionRule(models.Model):
+        "In code: 'var1 = record.default_code'",
+    )
+    description_template_type = fields.Selection(
+        [
+            ("plain", "Text"),
+            ("jinja", "Jinja"),
+            ("variables", "Variables"),
+        ],
```
Why do you need to add 3 different ways of rendering?
You could just rely on qweb, this is already usable with the current dependencies of this addon. 
Also I guess it would be fine to always try to render the description, IMO there is no need to declare it in a separate field. Just allow qweb syntax and render the description always.

mt-software-de replied:
Also Jinja2 is not really in use in newer Odoo Versions anymore.

raumschmiede-joshuaL replied:
> Why do you need to add 3 different ways of rendering?

For normal users Jinja is better. For devs that create the descriptions code is better. And plain just as default.



> always try to render the description

Depends. If we decide at the end to only have one for placeholders, then it is ok.



> Just allow qweb syntax and render the description always

QWeb is more for bigger templates to render. Feels weird to use this. More complex syntax for a small use-case.
But I will test this locally

> Also Jinja2 is not really in use in newer Odoo Versions anymore.

Don't know how many non-dev users would use Jinja. If the others are fine with it, I can remove it. For me ok

raumschmiede-joshuaL replied:
@mt-software-de, I implemented QWeb in a new commit for now. Looks like this: 
<img width="1389" height="581" alt="image" src="https://github.com/user-attachments/assets/7bcbd045-9692-41a4-bef4-0019bc5550ea" />

It feels more complicated to write this syntax. Jinja is almost the same, just with less wrapping around the actual code.
And for more complex code a separate code field is better.

Better to remove Jinja and Qweb

raumschmiede-joshuaL replied:
Removed QWeb and description code. Only Jinja and the description is always rendered

raumschmiede-joshuaL replied:
In the forward port of this PR for Odoo 16.0 I can replace Jinja with inline Qweb. Will check it after this here was merged

mt-software-de commented on base_exception/models/base_exception.py:
```diff
@@ -93,6 +221,13 @@ def _get_main_records(self):
     def _reverse_field(self):
         raise NotImplementedError()
 
+    def _get_sub_exception_field_names(self):
+        """
+        Used in case exceptions need to be checked on underlying records and the
+        detected exceptions need to be added to the parent record.
+        """
+        return []
+
```
Could you move those changes into an own commit?

raumschmiede-joshuaL replied:
Moved into PR https://github.com/OCA/server-tools/pull/3354

mt-software-de commented on base_exception/tests/purchase_test.py:
```diff
@@ -12,16 +13,27 @@ class ExceptionRule(models.Model):
     _inherit = "base.exception"
-    _name = "base.exception.test.purchase"
-    _description = "Base Exception Test Model"
+    _name = "test.purchase"
+    _description = "Test Model"
```
Is the change of `_name` necessary? This produces noise in the PR.

raumschmiede-joshuaL replied:
Not for `base.exception.test.purchase` but for the models below it. `base.exception.test.purchase.line.reverse.parent` is too long, when I started the unit test it said a table name for a relation was too long. That's why I shortened the names

raumschmiede-joshuaL replied:
Because of the backport of commit https://github.com/OCA/server-tools/pull/2670/commits/b5099bc51ea53592cefd5162b6ce72cc5e6edeee this change is not needed anymore. I removed it

mt-software-de commented on base_exception/tests/test_base_exception.py:
```diff
@@ -1,61 +1,26 @@
 
-    def test_valid(self):
+class TestBaseException(TestBaseExceptionCommon):
+    def test_01_valid(self):
```
Is changing the names of the test methods necessary? This produces noise in the PR.

raumschmiede-joshuaL replied:
Yes, to determine the order in which the tests are running. When debugging a test which runs as one of the last it is easier to deactivate the tests above that one because you know that they have a smaller number. Without numbers you have to check every test whether it comes in alphabetical order before or after your test you want to debug

mt-software-de replied:
There is no reason to do it, you could use test-tags to test only a specific unittest.

raumschmiede-joshuaL replied:
Clarified it. Moved this change into PR https://github.com/OCA/server-tools/pull/3354

mt-software-de reviewed (changes requested):
Thx. This sounds like a good extension. 

But there are a few things in need to mention. 

Could you please ensure to create separate commits for each functional change.

Also please try not to refactor the tests in the same commit as your functional changes. This creates a little bit of noise in this PR. Either remove those changes or create a separate commit for it.

mt-software-de commented:
IMO you should split this PR into 2.

One which is doing your changes for the `_get_sub_exception_field_names`, there you can also Backport https://github.com/OCA/server-tools/pull/2670/commits/b5099bc51ea53592cefd5162b6ce72cc5e6edeee from https://github.com/OCA/server-tools/pull/2670 because there are changes on the `_reverse_field`

And a second PR (this one)
Which just your updates for placeholders in description.
Here i would recommend only to use one method for rendering, since the inline_qweb is not available in v14 stick to `Jinja` without using the dependency to `mail` this would ease the FWP to newer versions.

hparfr reviewed (changes requested):
Hi @raumschmiede-joshuaL 

Thanks for your work on this module.

This module is quite sensible from a performance perspective; I fear the jinja rendering will put a penaltly. 

You may add this feature in a separate module or behind a feature flag.

raumschmiede-joshuaL commented:
@hparfr, thanks for your reply.

A flag like in the screenshot of the PR description? To choose between plain text description and Jinja. I can add this back

raumschmiede-joshuaL commented:
@hparfr, I can add it behind feature flag like in the 1st screenshot:

<img width="373" height="165" alt="image" src="https://github.com/user-attachments/assets/75f22a69-4005-4783-bc2f-e34e62d405d4" />


Can you check PR https://github.com/OCA/server-tools/pull/3354? Maybe that one can be merged 1st

raumschmiede-joshuaL committed:
[IMP] base_exception: Added _get_sub_exception_field_names

raumschmiede-joshuaL committed:
[IMP] base_exception: Added _detect_exceptions override
This ensures that the model on which the exceptions are detected get the exceptions assigned by default. Previously it was only possible if this def was overridden manually

raumschmiede-joshuaL committed:
[IMP] base_exception: Added Jinja placeholders to exception description

mt-software-de reviewed (changes requested):
IMO. We should get rid of the two different behaviors. That you can detect exceptions on subrecords and store it there and detect it on subrecords and not store it there. 

By storing the exceptions always where they are really detected, all of your changes could be simplified. 

Also IMO it would make sense to store the summary when the exceptions are detected.
From a performance aspect it should not make a huge difference for plain exceptions.

raumschmiede-joshuaL commented:
> IMO. We should get rid of the two different behaviors. That you can detect exceptions on subrecords and store it there and detect it on subrecords and not store it there.
> By storing the exceptions always where they are really detected, all of your changes could be simplified.


I think as there are 2 classes, `base.exception.method` and `base.exception`, 2 different behaviours are fine. For the case that you do not want to have `exception_ids` on a model where you will never use it.

 
> Also IMO it would make sense to store the summary when the exceptions are detected. From a performance aspect it should not make a huge difference for plain exceptions.

What is the advantage of saving the summary? The performance drain is probably not much but still a disadvantage for the current state. For my placeholders it is a disadvantage as data can change on a record which is used as placeholder value.
For example if `product_id.default_code` of a sales order line is changed but in the summary the old SKU is still shown until a recompute is triggered

vuwnevska commented on base_exception/models/base_exception.py:
```diff
@@ -91,6 +176,28 @@ def _popup_exceptions(self):
+        if not self._add_detected_exceptions_to_self():
+            return records
+
+        (self - records).exception_ids = [(3, rule.id)]
+        records.exception_ids = [(4, rule.id)]
```
IMO this is not obvious behaviour, and probably the deleted comment from detect_exceptions is valid here

raumschmiede-joshuaL replied:
If you have a field `exception_ids` on a model, you would expect that it will be set when exceptions are detected. Otherwise the field would be useless. If you do not want this behaviour, you would probably not want to have `exception_ids` on the record. For that you can inherit from `base.exception.method` instead of `base.exception`.


The way how the exceptions are assigned to the sub-records is the same as in `sale_exception`. From there I copied it. There the comment says that the flush optimizes it, In general I think the lock was maybe only possible because of the reverse field as a lot of writes could happen on the exception field, e.g. `sale_ids`. But this field is gone, so I guess there is no lock anymore
